x/ibchooks/marker_hooks.go

```diff
@@ -0,0 +1,226 @@
+package ibchooks
+
+import (
+	"encoding/json"
+	"strconv"
+	"strings"
+
+	sdkerrors "cosmossdk.io/errors"
+
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	transfertypes "github.com/cosmos/ibc-go/v6/modules/apps/transfer/types"
+	"github.com/cosmos/ibc-go/v6/modules/core/exported"
+	ibckeeper "github.com/cosmos/ibc-go/v6/modules/core/keeper"
+	tendermintclient "github.com/cosmos/ibc-go/v6/modules/light-clients/07-tendermint/types"
+
+	"github.com/provenance-io/provenance/x/ibchooks/types"
+	markerkeeper "github.com/provenance-io/provenance/x/marker/keeper"
+	markertypes "github.com/provenance-io/provenance/x/marker/types"
+)
+
+type MarkerHooks struct {
+	MarkerKeeper *markerkeeper.Keeper
+}
+
+func NewMarkerHooks(markerkeeper *markerkeeper.Keeper) MarkerHooks {
+	return MarkerHooks{
+		MarkerKeeper: markerkeeper,
+	}
+}
+
+// ProperlyConfigured returns false marker hooks are configured properly
+func (h MarkerHooks) ProperlyConfigured() bool {
+	return h.MarkerKeeper != nil
+}
+
+// AddUpdateMarker will add or update ibc Marker with transfer authorities
+func (h MarkerHooks) AddUpdateMarker(ctx sdktypes.Context, packet exported.PacketI, ibcKeeper *ibckeeper.Keeper) error {
+	var data transfertypes.FungibleTokenPacketData
+	if err := json.Unmarshal(packet.GetData(), &data); err != nil {
+		return err
+	}
+	ibcDenom := MustExtractDenomFromPacketOnRecv(packet)
+	if strings.HasPrefix(ibcDenom, "ibc/") {
+		markerAddress, err := markertypes.MarkerAddress(ibcDenom)
+		if err != nil {
+			return err
+		}
+		marker, err := h.MarkerKeeper.GetMarker(ctx, markerAddress)
+		if err != nil {
+			return err
+		}
+		var transferAuthAddrs []sdktypes.AccAddress
+		transferAuthAddrs, coinType, err := ProcessMarkerMemo(data.GetMemo())
+		if err != nil {
+			return err
+		}
+		if marker == nil {
+			amount, errParse := strconv.ParseInt(data.Amount, 10, 64)
+			if errParse != nil {
+				return errParse
+			}
+			marker = markertypes.NewMarkerAccount(
+				authtypes.NewBaseAccountWithAddress(markertypes.MustGetMarkerAddress(ibcDenom)),
+				sdktypes.NewInt64Coin(ibcDenom, amount),
+				nil,
+				nil,
+				markertypes.StatusActive,
+				coinType,
+				false, // supply fixed
+				false, // allow gov
+				false, // allow force transfer
+				[]string{},
+			)
```

Comment:
The logic in here needs to be discussed and understood by the team.  I took a best first guess at how these should be setup, but decided to defer it to a larger discussion.

Reply:
This logic makes sense to me. I haven't looked at the send portion yet.